hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hts/airdrops/TokenAirdropDecoder.java

```diff
@@ -0,0 +1,126 @@
+import com.hedera.hapi.node.base.TokenTransferList;
+import com.hedera.hapi.node.token.TokenAirdropTransactionBody;
+import com.hedera.hapi.node.transaction.TransactionBody;
+import com.hedera.node.app.service.contract.impl.exec.systemcontracts.hts.AddressIdConverter;
+import com.hedera.node.app.service.contract.impl.exec.systemcontracts.hts.HtsCallAttempt;
+import com.hedera.node.app.service.contract.impl.utils.ConversionUtils;
+import com.hedera.node.config.data.LedgerConfig;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.util.ArrayList;
+import java.util.Arrays;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+
+@Singleton
+public class TokenAirdropDecoder {
+
+    // Tuple indexes
+    private static final int TOKEN = 0;
+    private static final int TOKEN_TRANSFERS = 1;
+    private static final int NFT_AMOUNT = 2;
+    private static final int TOKEN_ACCOUNT_ID = 0;
+    private static final int TOKEN_AMOUNT = 1;
+    private static final int TOKEN_IS_APPROVAL = 2;
+    private static final int NFT_SENDER = 0;
+    private static final int NFT_RECEIVER = 1;
+    private static final int NFT_SERIAL = 2;
+    private static final int NFT_IS_APPROVAL = 3;
+
+    @Inject
+    public TokenAirdropDecoder() {
+        // Dagger2
+    }
+
+    public TransactionBody decodeAirdrop(@NonNull final HtsCallAttempt attempt) {
+        final var call = TokenAirdropTranslator.TOKEN_AIRDROP.decodeCall(attempt.inputBytes());
+        final var transferList = (Tuple[]) call.get(0);
+        final var ledgerConfig = attempt.configuration().getConfigData(LedgerConfig.class);
+        final var tokenAirdrop = bodyForAirdrop(transferList, attempt.addressIdConverter(), ledgerConfig);
+        return TransactionBody.newBuilder().tokenAirdrop(tokenAirdrop).build();
+    }
+
+    private TokenAirdropTransactionBody bodyForAirdrop(
+            @NonNull final Tuple[] transferList,
+            @NonNull final AddressIdConverter addressIdConverter,
+            @NonNull final LedgerConfig ledgerConfig) {
+        final var transferBuilderList = new ArrayList<TokenTransferList>();
+        validateSemantics(transferList, ledgerConfig);
+        Arrays.stream(transferList).forEach(transfer -> {
+            final var tokenTransferList = TokenTransferList.newBuilder();
+            final var token = ConversionUtils.asTokenId(transfer.get(TOKEN));
+            tokenTransferList.token(token);
+            final var tokenAmountsTuple = (Tuple[]) transfer.get(TOKEN_TRANSFERS);
+            final var nftAmountsTuple = (Tuple[]) transfer.get(NFT_AMOUNT);
+            if (tokenAmountsTuple.length > 0) {
+                final var aaList = new ArrayList<AccountAmount>();
+                Arrays.stream(tokenAmountsTuple).forEach(tokenAmount -> {
+                    final var amount = (long) tokenAmount.get(TOKEN_AMOUNT);
+                    final var account = addressIdConverter.convert(tokenAmount.get(TOKEN_ACCOUNT_ID));
```

Comment:
If you look at `ClassicTransfersCall.execute`, there is a check to see if we are trying to transfer to a system account.  Do we have/want a similar check here?

Reply:
We have that check already in the Handler validations. Do you think we should include it here as well, just in case?

Reply:
I don't think it could hurt.  Can be done later at your discretion.

Reply:
Added the check.